src/Container/Tarball.hs

```diff
@@ -0,0 +1,183 @@
+{-# LANGUAGE BangPatterns #-}
+
+module Container.Tarball (
+  -- * for testing
+  TarEntries (..),
+  removeWhiteOut,
+  mkEntries,
+  mkImage,
+) where
+
+import Codec.Archive.Tar (
+  Entry (entryContent),
+  EntryContent (NormalFile, SymbolicLink),
+  entryPath,
+ )
+import Codec.Archive.Tar qualified as Tar
+import Codec.Archive.Tar.Entry (Entry (entryTarPath), TarPath, fromTarPath)
+import Codec.Archive.Tar.Entry qualified as TarEntry
+import Codec.Archive.Tar.Index (TarEntryOffset, nextEntryOffset)
+import Container.Errors (
+  ContainerImgParsingError (
+    ContainerNoLayersDiscovered,
+    TarLayerNotAFile,
+    TarMissingLayerTar,
+    TarParserError
+  ),
+ )
+import Container.Types (
+  ContainerFSChangeSet (InsertOrUpdate, Whiteout),
+  ContainerImageRaw (ContainerImageRaw),
+  ContainerLayer (
+    ContainerLayer,
+    lastOffset,
+    layerChangeSets
+  ),
+ )
+import Data.Either (lefts, rights)
+import Data.List.NonEmpty qualified as NLE
+import Data.Sequence (Seq, ViewL (EmptyL, (:<)), viewl, (|>))
+import Data.Sequence qualified as Seq
+import Data.String.Conversion (ToText (toText), toString)
+import Data.Text (Text)
+import Data.Text qualified as Text
+
+data TarEntries = TarEntries
+  { entries :: !(Seq (Tar.Entry, TarEntryOffset))
+  , prevOffset :: !TarEntryOffset
+  }
+  deriving (Show)
+
+mkEntries ::
+  Tar.Entries Tar.FormatError ->
+  Either ContainerImgParsingError TarEntries
+mkEntries = build (TarEntries mempty 0)
+  where
+    build :: TarEntries -> Tar.Entries Tar.FormatError -> Either ContainerImgParsingError TarEntries
+    build !builder (Tar.Next e es) = build (addEntry e builder) es
+    build !builder Tar.Done = Right builder
+    build !_ (Tar.Fail err) = Left $ TarParserError err
+
+    addEntry :: Tar.Entry -> TarEntries -> TarEntries
+    addEntry entry tarEntries =
+      tarEntries
+        { entries = entries tarEntries |> (entry, prevOffset tarEntries)
+        , prevOffset = nextEntryOffset entry (prevOffset tarEntries)
+        }
+
+{- HLINT ignore "Avoid restricted function" -}
+mkImage ::
+  TarEntries ->
+  NLE.NonEmpty FilePath ->
+  Either (NLE.NonEmpty ContainerImgParsingError) ContainerImageRaw
+mkImage entries layerTarballPaths =
+  if null errs
+    then case parsedLayers of
+      [] -> Left $ NLE.fromList [ContainerNoLayersDiscovered]
+      l -> Right $ ContainerImageRaw $ NLE.fromList l
+    else -- This is safe, since we asserted that layers are not null
+    -- earlier in main clause.
+      Left $ NLE.fromList errs
```

Comment:
It's possible to avoid the partial functions here using pattern matches. I've compressed the two layers of logic here into one using a tuple, but if it's easier to read it can be converted to nested `case`s too.

```suggestion
mkImage ::
  TarEntries ->
  NLE.NonEmpty FilePath ->
  Either (NLE.NonEmpty ContainerImgParsingError) ContainerImageRaw
mkImage entries layerTarballPaths =
  case (errs, parsedLayers) of
    ((e : es), _) -> Left $ e NLE.:| es
    (_, []) -> Left $ NLE.singleton ContainerNoLayersDiscovered
    (_, (l : ls)) -> Right . ContainerImageRaw $ l NLE.:| ls
```